ton_lib_core/src/traits/metadata.rs

```diff
@@ -0,0 +1,12 @@
+use crate::cell::TonHash;
+use crate::error::TLCoreError;
+use std::collections::HashMap;
+pub trait Metadata: Sized {
+    fn from_data(
+        onchain: Option<&HashMap<TonHash, impl AsRef<[u8]>>>,
+        offchain: Option<&str>,
+    ) -> Result<Self, TLCoreError>;
+
+    fn from_offchain(offchain: &str) -> Result<Self, TLCoreError>;
```

Comment:
Instead of implementing it for all types, provide a default implementation:

```
Self::from_data(None, Some(offchain)
```

And the same for the second one

But I would rename the functions to `from_json` and `from_dict`